src/lib/utils/lossless_predicate_cast.cpp

```diff
@@ -0,0 +1,63 @@
+#include "lossless_predicate_cast.hpp"
+
+#include <cmath>
+
+namespace opossum {
+
+std::optional<float> next_float_towards(const double value, const double towards) {
+  // See lossless_cast<float>(double)
+  if (value > 340282346638528859811704183484516925440.0 || value < -340282346638528859811704183484516925440.0) {
+    return std::nullopt;
+  }
+
+  if (value == towards) return std::nullopt;
+
+  const auto casted_value = static_cast<float>(value);
+
+  if (casted_value < value && towards < value) return casted_value;
+  if (casted_value > value && towards > value) return casted_value;
+
+  const float next = std::nexttowardf(casted_value, towards);
+
+  // Maybe someone smarter understands all the edge cases of floats. I'd rather be on the safe side.
+  if (!std::isfinite(next)) return std::nullopt;
+
+  return next;
+}
+
+std::optional<std::pair<PredicateCondition, AllTypeVariant>> lossless_predicate_variant_cast(
+    const PredicateCondition condition, const AllTypeVariant& variant, DataType target_data_type) {
+  // Code duplication with lossless_cast.cpp, but it's are already hard enough to read
+
+  const auto source_data_type = data_type_from_all_type_variant(variant);
+
+  // Lossless casting from NULL to NULL is always NULL. (Cannot be handled below as resolve_data_type()
+  // doesn't resolve NULL)
+  if (source_data_type == DataType::Null && target_data_type == DataType::Null) {
+    // We should be able to return {condition, variant} here, but clang-tidy has what I believe to be a false
+    // positive. Instead of ignoring it and risking overlooking something, we let the ExpressionEvaluator handle
+    // this case. This is only for stupid queries like `WHERE x = NULL`, anyway.
+    return std::nullopt;
+  }
+
+  // Safe casting between NULL and non-NULL type is not possible. (Cannot be handled below as resolve_data_type()
+  // doesn't resolve NULL)
+  if ((source_data_type == DataType::Null) != (target_data_type == DataType::Null)) {
+    return std::nullopt;
+  }
+
```

Comment:
This could also be one if-clause with `source_data_type == DataType::Null || target_data_type == DataType::Null`

Reply:
Do you mean combining this with the clause above? That might logically be the case, but my idea was to keep the two cases (that came from lossless_predicate) separate. I hoped this would make them easier to understand - what do you think?

Reply:
I was about to submit that when you changed the code. I liked your previous version

Reply:
... now what? :D 
Can someone come up with a compromise?

Reply:
After discussing this with @Bensk1, we have agreed to go back to the original version. One reason being the similarity to lossless_cast, the other one being that these are two distinct cases. Nonetheless, I appreciate the feedback!